lib/AST/SubstitutionMap.cpp

```diff
@@ -684,13 +684,23 @@ void SubstitutionMap::profile(llvm::FoldingSetNodeID &id) const {
 }
 
 bool SubstitutionMap::isIdentity() const {
-  for (unsigned i : indices(getReplacementTypes())) {
-    auto replacement = dyn_cast<GenericTypeParamType>(
-      getReplacementTypes()[i]->getCanonicalType(getGenericSignature()));
-    if (!replacement)
-      return false;
-    if (getGenericSignature()->getGenericParamOrdinal(replacement) != i)
-      return false;
-  }
-  return true;
+  GenericSignature *sig = getGenericSignature();
+  unsigned countOfGenericParams = 0;
+  bool hasNonIdentityReplacement = false;
+
+  sig->forEachParam([&](GenericTypeParamType *paramTy, bool isCanonical) {
+    ++countOfGenericParams;
+    if (!isCanonical)
+      return;
+
+    unsigned ordinal = sig->getGenericParamOrdinal(paramTy);
+    auto replacement = getReplacementTypes()[ordinal]->getCanonicalType(sig);
+    if (!isa<GenericTypeParamType>(replacement) ||
+        replacement != paramTy->getCanonicalType()) {
+      hasNonIdentityReplacement = true;
+    }
+  });
+
+  return !hasNonIdentityReplacement &&
+         countOfGenericParams == getReplacementTypes().size();
```

Comment:
Is it ever possible for countOfGenericParams to not equal getReplacementTypes().size()? Maybe it should be an assertion instead?

Reply:
I was wondering that, but I didn't know if it was safe. All valid SubstitutionMaps should behave like that, but what about a degenerate one?

Reply:
See above :)